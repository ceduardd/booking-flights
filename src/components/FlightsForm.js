import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import FormTabs from './FormTabs';
import TypeFlightSelector from './selectors/TypeFlightSelector';
import PlaceSelector from './selectors/PlaceSelector';
import DateSelector from './selectors/DateSelector';
import ClassSelector from './selectors/ClassSelector';
import TravelerSelector from './selectors/TravelerSelector';
import ChildrenIcon from './icons/ChildrenIcon';
import InfantIcon from './icons/InfantIcon';
import AdultIcon from './icons/AdultIcon';
import ErrorLabel from './ErrorLabel';
import {
  removeErrorDate,
  removeErrorFormat,
  removeWrongDates,
  removeWrongPlaces,
  showErrorDate,
  showErrorFormat,
  showWrongDates,
  showWrongPlaces,
} from '../actions/uiActions';

// Initial travel dates
const now = moment().minutes(0).seconds(0).add(1, 'hours'); // 3:00:00
const nowPlus3 = now.clone().add(3, 'd');

const FlightsForm = () => {
  const dispatch = useDispatch();

  // Get realted inputs errors
  const { wrongDatesError, wrongPlacesError } = useSelector(state => state.ui);

  // Track inputs state
  const [origin, setOrigin] = useState('');
  const [destination, setDestination] = useState('');
  const [outboundDate, setOutboundDate] = useState(now.toDate());
  const [returnDate, setReturnDate] = useState(nowPlus3.toDate());
  const [adults, setAdults] = useState(1);
  const [children, setChildren] = useState(0);
  const [infants, setInfants] = useState(0);
  const [classFlight, setClassFlight] = useState('Economy');
  const [flightType, setFlightType] = useState('Round Trip');

  const clearErrors = () => {
    dispatch(removeErrorFormat());
    dispatch(removeErrorDate());
    dispatch(removeWrongDates());
    dispatch(removeWrongPlaces());
  };

  const submitHandler = e => {
    e.preventDefault();

    // Before new checking
    clearErrors();

    // Get places entries to check
    const from = origin.split(' ');
    const to = destination.split(' ');

    if (origin === destination) {
      dispatch(
        showWrongPlaces('The departure and arrival airports cannot be the same')
      );
    } else if (from.length <= 1) {
      dispatch(showErrorFormat('Select an available origin'));
    } else if (to.length <= 1) {
      dispatch(showErrorFormat('Select an available destination'));
    } else if (outboundDate <= moment.now()) {
      dispatch(showErrorDate('Outbound date is invalid'));
    } else if (returnDate <= moment.now()) {
      dispatch(showErrorDate('Return date is invalid'));
    } else if (returnDate < outboundDate) {
      dispatch(showWrongDates('Wrong dates order'));
    } else {
      const originCode = from[from.length - 1];
      const destinationCode = to[to.length - 1];

      // Build URI to deep linking
      const URI = `https://www.swiss.com/us/en/Book/Outbound/${originCode}-${destinationCode}/from-${moment(
        outboundDate
      ).format(
        'YYYY-MM-DD'
      )}/adults-${adults}/children-${children}/infants-${infants}/class-${classFlight.toLowerCase()}/al-LX/sidmbvl`;

      // Search with SWISS on new tab
      window.open(URI);
    }
  };

  return (
    <div className="home__form-container">
      <FormTabs />

      <form
        className="home__form-entries"
        autoComplete="off"
        onSubmit={submitHandler}
      >
        <TypeFlightSelector value={flightType} changeHandler={setFlightType} />

        <div className="home__places">
          {wrongPlacesError && <ErrorLabel>{wrongPlacesError}</ErrorLabel>}

          <PlaceSelector
            label="From"
            value={origin}
            changeHandler={setOrigin}
            placeholder="Enter your origin"
            menuInstruction="Select an origin"
            errorMessage="Select an available origin"
          />

          <PlaceSelector
            label="To"
            value={destination}
            changeHandler={setDestination}
            placeholder="Enter your destination"
            menuInstruction="Select a destination"
            errorMessage="Select an available destination"
          />
        </div>

        <div className="home__form-dates">
          {wrongDatesError && <ErrorLabel>{wrongDatesError}</ErrorLabel>}

          <DateSelector
            label="Outbound"
            value={outboundDate}
            changeHandler={setOutboundDate}
            errorMessage="Outbound date is invalid"
          />

          {flightType !== 'One Way' && (
            <DateSelector
              label="Return"
              value={returnDate}
              changeHandler={setReturnDate}
              disabled={flightType !== 'Round Trip'}
              errorMessage="Return date is invalid"
            />
          )}
        </div>

        <ClassSelector
          label="Class"
          placeholder="Seleact a class flight"
          value={classFlight}
          changeHandler={setClassFlight}
          menuInstruction="Seleact a class flight"
        />

        <div className="home__form-travelers">
          <TravelerSelector
            label="Adutls"
            value={adults}
            changeHandler={setAdults}
            icon={AdultIcon}
            min={0}
          />
          <TravelerSelector
            label="Children"
            value={children}
            changeHandler={setChildren}
            icon={ChildrenIcon}
            min={0}
          />
          <TravelerSelector
            label="Infants"
            value={infants}
            changeHandler={setInfants}
            icon={InfantIcon}
            min={0}
          />
        </div>

        {/* Submit btn */}
        <button className="home__form-btn" type="submit">
          Search
        </button>
      </form>
    </div>
  );
};

export default FlightsForm;
