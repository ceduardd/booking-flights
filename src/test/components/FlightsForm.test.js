import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import 'jest-canvas-mock';

import FlightsForm from '../../components/FlightsForm';

const mockStore = configureStore();

const initialState = {
  ui: {},
};

let store = mockStore(initialState);

store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <FlightsForm />
  </Provider>
);

describe('<FlightsForm />', () => {
  beforeEach(() => {
    store = mockStore(initialState);
    jest.clearAllMocks();
  });

  test('should be match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should show something error', () => {
    const initialState = {
      ui: {
        wrongDatesError: 'Select an available origin',
      },
    };

    const store = mockStore(initialState);

    store.dispatch = jest.fn();

    const wrapper = mount(
      <Provider store={store}>
        <FlightsForm />
      </Provider>
    );

    const form = wrapper.find('form');

    form.simulate('submit', { pereventDefault() {} });

    expect(wrapper.find('ErrorLabel').exists()).toBe(true);
  });
});
