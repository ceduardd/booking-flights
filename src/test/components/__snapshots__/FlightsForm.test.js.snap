// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlightsForm /> should be match with snapshot 1`] = `
<div
  className="home__form-container"
>
  <div
    className="home__form-tabs"
  >
    <button
      className="active"
      type="button"
    >
      <i
        className="fas fa-plane"
      />
       Flights
    </button>
    <button
      type="button"
    >
      <i
        className="fas fa-car"
      />
       Cars
    </button>
    <button
      type="button"
    >
      <i
        className="fas fa-hotel"
      />
       Hotel
    </button>
  </div>
  <form
    autoComplete="off"
    className="home__form-entries"
    onSubmit={[Function]}
  >
    <div
      className="home__form-flight-type"
    >
      <label
        className="home__form-radio-btn not-active"
        htmlFor="oneWay"
      >
        One Way
        <input
          checked={false}
          id="oneWay"
          name="flightType"
          onChange={[Function]}
          type="radio"
          value="One Way"
        />
      </label>
      <label
        className="home__form-radio-btn active"
        htmlFor="roundTrip"
      >
        Round Trip
        <input
          checked={true}
          id="roundTrip"
          name="flightType"
          onChange={[Function]}
          type="radio"
          value="Round Trip"
        />
      </label>
    </div>
    <div
      className="home__places"
    >
      <div
        className="selector__container"
      >
        <label>
          <span>
            From
            :
          </span>
          <div
            className="form-group"
            style={
              Object {
                "marginTop": "5px",
              }
            }
          >
            <input
              className="form-control"
              onChange={[Function]}
              placeholder="Enter your origin"
              type="text"
              value=""
            />
            <i
              className="fas fa-map-marker-alt"
            />
          </div>
        </label>
      </div>
      <div
        className="selector__container"
      >
        <label>
          <span>
            To
            :
          </span>
          <div
            className="form-group"
            style={
              Object {
                "marginTop": "5px",
              }
            }
          >
            <input
              className="form-control"
              onChange={[Function]}
              placeholder="Enter your destination"
              type="text"
              value=""
            />
            <i
              className="fas fa-map-marker-alt"
            />
          </div>
        </label>
      </div>
    </div>
    <div
      className="home__form-dates"
    >
      <label
        className="date-selector"
        htmlFor="from"
      >
        <span>
          Outbound
          :
        </span>
        <div
          className="form-group"
          style={
            Object {
              "marginTop": "5px",
            }
          }
        >
          <div
            className="react-date-picker react-date-picker--closed react-date-picker--enabled"
            onFocus={[Function]}
          >
            <div
              className="react-date-picker__wrapper"
            >
              <div
                className="react-date-picker__inputGroup"
                onClick={[Function]}
              >
                <input
                  max="275760-09-12"
                  min="0000-12-31"
                  name="date"
                  onChange={[Function]}
                  onFocus={[Function]}
                  style={
                    Object {
                      "position": "absolute",
                      "visibility": "hidden",
                      "zIndex": "-999",
                    }
                  }
                  type="date"
                  value="2021-01-27"
                />
                <input
                  autoComplete="off"
                  className="react-date-picker__inputGroup__input react-date-picker__inputGroup__month"
                  data-input="true"
                  inputMode="numeric"
                  key="input"
                  max={12}
                  min={1}
                  name="month"
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyPress={[Function]}
                  onKeyUp={[Function]}
                  placeholder="--"
                  required={null}
                  type="number"
                  value={1}
                />
                <span
                  className="react-date-picker__inputGroup__divider"
                >
                  /
                </span>
                <input
                  autoComplete="off"
                  autoFocus={false}
                  className="react-date-picker__inputGroup__input react-date-picker__inputGroup__day"
                  data-input="true"
                  inputMode="numeric"
                  key="input"
                  max={31}
                  min={1}
                  name="day"
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyPress={[Function]}
                  onKeyUp={[Function]}
                  placeholder="--"
                  required={null}
                  type="number"
                  value={27}
                />
                <span
                  className="react-date-picker__inputGroup__divider"
                >
                  /
                </span>
                <input
                  autoComplete="off"
                  autoFocus={false}
                  className="react-date-picker__inputGroup__input react-date-picker__inputGroup__year"
                  data-input="true"
                  inputMode="numeric"
                  key="input"
                  max={275760}
                  min={1}
                  name="year"
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyPress={[Function]}
                  onKeyUp={[Function]}
                  placeholder="----"
                  required={null}
                  step={1}
                  type="number"
                  value={2021}
                />
              </div>
              <button
                className="react-date-picker__calendar-button react-date-picker__button"
                onClick={[Function]}
                onFocus={[Function]}
                type="button"
              >
                <i
                  className="fas fa-chevron-down"
                />
              </button>
            </div>
          </div>
          <i
            className="far fa-calendar"
          />
        </div>
      </label>
      <label
        className="date-selector"
        htmlFor="from"
      >
        <span>
          Return
          :
        </span>
        <div
          className="form-group"
          style={
            Object {
              "marginTop": "5px",
            }
          }
        >
          <div
            className="react-date-picker react-date-picker--closed react-date-picker--enabled"
            onFocus={[Function]}
          >
            <div
              className="react-date-picker__wrapper"
            >
              <div
                className="react-date-picker__inputGroup"
                onClick={[Function]}
              >
                <input
                  disabled={false}
                  max="275760-09-12"
                  min="0000-12-31"
                  name="date"
                  onChange={[Function]}
                  onFocus={[Function]}
                  style={
                    Object {
                      "position": "absolute",
                      "visibility": "hidden",
                      "zIndex": "-999",
                    }
                  }
                  type="date"
                  value="2021-01-30"
                />
                <input
                  autoComplete="off"
                  className="react-date-picker__inputGroup__input react-date-picker__inputGroup__month"
                  data-input="true"
                  disabled={false}
                  inputMode="numeric"
                  key="input"
                  max={12}
                  min={1}
                  name="month"
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyPress={[Function]}
                  onKeyUp={[Function]}
                  placeholder="--"
                  required={null}
                  type="number"
                  value={1}
                />
                <span
                  className="react-date-picker__inputGroup__divider"
                >
                  /
                </span>
                <input
                  autoComplete="off"
                  autoFocus={false}
                  className="react-date-picker__inputGroup__input react-date-picker__inputGroup__day"
                  data-input="true"
                  disabled={false}
                  inputMode="numeric"
                  key="input"
                  max={31}
                  min={1}
                  name="day"
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyPress={[Function]}
                  onKeyUp={[Function]}
                  placeholder="--"
                  required={null}
                  type="number"
                  value={30}
                />
                <span
                  className="react-date-picker__inputGroup__divider"
                >
                  /
                </span>
                <input
                  autoComplete="off"
                  autoFocus={false}
                  className="react-date-picker__inputGroup__input react-date-picker__inputGroup__year"
                  data-input="true"
                  disabled={false}
                  inputMode="numeric"
                  key="input"
                  max={275760}
                  min={1}
                  name="year"
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyPress={[Function]}
                  onKeyUp={[Function]}
                  placeholder="----"
                  required={null}
                  step={1}
                  type="number"
                  value={2021}
                />
              </div>
              <button
                className="react-date-picker__calendar-button react-date-picker__button"
                disabled={false}
                onClick={[Function]}
                onFocus={[Function]}
                type="button"
              >
                <i
                  className="fas fa-chevron-down"
                />
              </button>
            </div>
          </div>
          <i
            className="far fa-calendar"
          />
        </div>
      </label>
    </div>
    <div
      className="selector__container"
    >
      <label>
        <span>
          Class
          :
        </span>
        <div
          className="form-group"
          style={
            Object {
              "marginTop": "5px",
            }
          }
        >
          <input
            className="form-control disabled"
            disabled={true}
            onChange={[Function]}
            placeholder="Seleact a class flight"
            type="text"
            value="Economy"
          />
          <i
            className="far fa-star"
          />
          <i
            className="fas fa-chevron-down dropdown-indicator"
            onClick={[Function]}
          />
        </div>
      </label>
    </div>
    <div
      className="home__form-travelers"
    >
      <label
        className="traveler-selector"
      >
        <span>
          Adutls
          :
        </span>
        <div
          className="form-group"
          style={
            Object {
              "marginTop": "5px",
            }
          }
        >
          <input
            className="form-control"
            min={0}
            onChange={[Function]}
            type="number"
            value={1}
          />
          <i
            className="fas fa-male"
          />
        </div>
      </label>
      <label
        className="traveler-selector"
      >
        <span>
          Children
          :
        </span>
        <div
          className="form-group"
          style={
            Object {
              "marginTop": "5px",
            }
          }
        >
          <input
            className="form-control"
            min={0}
            onChange={[Function]}
            type="number"
            value={0}
          />
          <i
            className="fas fa-child"
          />
        </div>
      </label>
      <label
        className="traveler-selector"
      >
        <span>
          Infants
          :
        </span>
        <div
          className="form-group"
          style={
            Object {
              "marginTop": "5px",
            }
          }
        >
          <input
            className="form-control"
            min={0}
            onChange={[Function]}
            type="number"
            value={0}
          />
          <i
            className="fas fa-baby"
          />
        </div>
      </label>
    </div>
    <button
      className="home__form-btn"
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;
